File: backend/app/services/bookmark_service.py
from typing import List, Optional

from fastapi import HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models import Bookmark, Question
from app.schemas.bookmark import BookmarkWithQuestion


async def add_bookmark(db: AsyncSession, user_id: int, question_id: int) -> Bookmark:
    # Check question exists
    q_result = await db.execute(select(Question).where(Question.id == question_id))
    if not q_result.scalar_one_or_none():
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Question not found")

    # Check if already bookmarked
    result = await db.execute(
        select(Bookmark).where(Bookmark.user_id == user_id, Bookmark.question_id == question_id)
    )
    existing = result.scalar_one_or_none()
    if existing:
        return existing

    bookmark = Bookmark(user_id=user_id, question_id=question_id)
    db.add(bookmark)
    await db.commit()
    await db.refresh(bookmark)
    return bookmark


async def remove_bookmark(db: AsyncSession, user_id: int, question_id: int) -> None:
    result = await db.execute(
        select(Bookmark).where(Bookmark.user_id == user_id, Bookmark.question_id == question_id)
    )
    bookmark = result.scalar_one_or_none()
    if not bookmark:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Bookmark not found")

    await db.delete(bookmark)
    await db.commit()


async def get_bookmarks(
    db: AsyncSession, user_id: int, exam_id: Optional[int] = None
) -> List[BookmarkWithQuestion]:
    query = (
        select(Bookmark, Question)
        .join(Question, Bookmark.question_id == Question.id)
        .where(Bookmark.user_id == user_id)
    )
    if exam_id:
        query = query.where(Question.exam_id == exam_id)
    query = query.order_by(Bookmark.created_at.desc())

    result = await db.execute(query)
    items = []
    for bm, q in result.all():
        items.append(
            BookmarkWithQuestion(
                id=bm.id,
                question_id=bm.question_id,
                created_at=bm.created_at,
                question_text=q.question_text,
                question_type=q.question_type,
                exam_id=q.exam_id,
            )
        )
    return items

File: workers-api/src/services/bookmark-service.ts
import { eq, and, desc } from "drizzle-orm";
import type { Database } from "../db/client";
import { bookmarks, questions } from "../db/schema";
import { AppError } from "../lib/errors";

export async function addBookmark(db: Database, userId: number, questionId: number) {
  // Verify question exists
  const question = await db.query.questions.findFirst({
    where: eq(questions.id, questionId),
  });
  if (!question) throw new AppError(404, "Question not found");

  // Check if already bookmarked
  const existing = await db.query.bookmarks.findFirst({
    where: and(eq(bookmarks.userId, userId), eq(bookmarks.questionId, questionId)),
  });
  if (existing) {
    return {
      id: existing.id,
      question_id: existing.questionId,
      created_at: existing.createdAt,
    };
  }

  const [created] = await db
    .insert(bookmarks)
    .values({ userId, questionId })
    .returning();

  return {
    id: created.id,
    question_id: created.questionId,
    created_at: created.createdAt,
  };
}

export async function removeBookmark(db: Database, userId: number, questionId: number) {
  const existing = await db.query.bookmarks.findFirst({
    where: and(eq(bookmarks.userId, userId), eq(bookmarks.questionId, questionId)),
  });
  if (!existing) throw new AppError(404, "Bookmark not found");

  await db.delete(bookmarks).where(eq(bookmarks.id, existing.id));
}

export async function getBookmarks(
  db: Database,
  userId: number,
  examId?: number
) {
  let condition = eq(bookmarks.userId, userId);

  const rows = await db
    .select({
      id: bookmarks.id,
      questionId: bookmarks.questionId,
      createdAt: bookmarks.createdAt,
      questionText: questions.questionText,
      questionType: questions.questionType,
      examId: questions.examId,
    })
    .from(bookmarks)
    .innerJoin(questions, eq(bookmarks.questionId, questions.id))
    .where(
      examId
        ? and(condition, eq(questions.examId, examId))
        : condition
    )
    .orderBy(desc(bookmarks.createdAt));

  return rows.map((r) => ({
    id: r.id,
    question_id: r.questionId,
    created_at: r.createdAt,
    question_text: r.questionText,
    question_type: r.questionType,
    exam_id: r.examId,
  }));
}
